airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java

```diff
@@ -455,17 +462,91 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(final
+      final List<ConfiguredAirbyteStream> streamsForCtidSync = new ArrayList<>();
+      catalog.getStreams().stream()
+          .filter(stream -> streamsStillInCtidSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
+          .map(Jsons::clone)
+          .forEach(streamsForCtidSync::add);
+
+      streamsForCtidSync.addAll(newlyAddedStreams);
+
+      final List<ConfiguredAirbyteStream> streamsForXminSync = catalog.getStreams().stream()
+          .filter(stream -> !streamsForCtidSync.contains(stream))
+          .map(Jsons::clone)
+          .toList();
+
+      LOGGER.info("Streams to be synced via ctid : {}", streamsForCtidSync.size());
+      LOGGER.info("Streams to be synced via xmin : {}", streamsForXminSync.size());
+
+      final List<AutoCloseableIterator<AirbyteMessage>> ctidIterator = new ArrayList<>();
+      final List<AutoCloseableIterator<AirbyteMessage>> xminIterator = new ArrayList<>();
+
+      if (!streamsForCtidSync.isEmpty()) {
+        final CtidStateManager ctidStateManager = new CtidStateManager(statesFromCtidSync);
+        final PostgresCtidHandler ctidHandler = new PostgresCtidHandler(database, sourceOperations, getQuoteString(), ctidStateManager,
+            x -> new ObjectMapper().valueToTree(xminStatus),
+            (pair, jsonState) -> XminStateManager.getAirbyteStateMessage(pair, Jsons.object(jsonState, XminStatus.class)));
+        ctidIterator.addAll(ctidHandler.getIncrementalIterators(
+            new ConfiguredAirbyteCatalog().withStreams(streamsForCtidSync), tableNameToTable, emittedAt));
+      }
+
+      if (!streamsForXminSync.isEmpty()) {
+        final XminStateManager xminStateManager = new XminStateManager(statesFromXminSync);
+        final PostgresXminHandler xminHandler = new PostgresXminHandler(database, sourceOperations, getQuoteString(), xminStatus, xminStateManager);
+
+        xminIterator.addAll(xminHandler.getIncrementalIterators(
+            new ConfiguredAirbyteCatalog().withStreams(streamsForXminSync), tableNameToTable, emittedAt));
+      }
+
+      return Stream
+          .of(ctidIterator, xminIterator)
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
     } else {
-      final CtidStateManager ctidStateManager = new CtidStateManager(stateManager.getRawStateMessages());
-      final PostgresCtidHandler handler = new PostgresCtidHandler(database, sourceOperations, getQuoteString(), null, ctidStateManager);
-      return handler.getIncrementalIterators(catalog, tableNameToTable, emittedAt);
-//      return super.getIncrementalIterators(database, catalog, tableNameToTable, stateManager, emittedAt);
+      return super.getIncrementalIterators(database, catalog, tableNameToTable, stateManager, emittedAt);
     }
   }
 
+  protected List<ConfiguredAirbyteStream> identifyNewlyAddedStreams(final ConfiguredAirbyteCatalog catalog, final Set<AirbyteStreamNameNamespacePair> alreadySeenStreams) {
```

Comment:
nit : this and the others can we move to a utility class instead of `PostgresSource`?